src/llm/servable_initializer.cpp

```diff
@@ -49,11 +51,106 @@
     py::gil_scoped_acquire acquire;
     try {
-        auto locals = py::dict("templates_directory"_a = chatTemplateDirectory);
+        bool isGGUFModel = checkIfGGUFModel(chatTemplateDirectory);
+        // we need to pass tokenizer template and bos/eos tokens to python code
+        // if we have GGUF model, we will use them to create a template object
+        std::string tokenizerTemplate;
+        std::string tokenizerBosToken;
+        std::string tokenizerEosToken;
+        if (isGGUFModel) {
+            tokenizerTemplate = properties->tokenizer.get_chat_template();
+            tokenizerBosToken = properties->tokenizer.get_bos_token();
+            tokenizerEosToken = properties->tokenizer.get_eos_token();
+
+            // Workaround for CVS-172426
+            if (tokenizerBosToken.empty() || tokenizerEosToken.empty()) {
+                // time measure following if statement
+                std::chrono::steady_clock::time_point start = std::chrono::steady_clock::now();
+                // if tokenizer bos/eos tokens are empty, we will try to get them from tokenizer vocab
+                std::pair<std::optional<std::string>, std::optional<std::string>> tokens;
+                tokens = getBosAndEosTokenFromTokenizerVocab(properties->tokenizer);
+                if (tokens.first.has_value()) {
+                    tokenizerBosToken = tokens.first.value();
+                }
+                if (tokens.second.has_value()) {
+                    tokenizerEosToken = tokens.second.value();
+                }
+                SPDLOG_TRACE("Tokenizer bos token: {}, eos token: {}, bos token id: {}, eos token id: {} isGGUF:{} chat_template from tokenizer: \n{}",
+                    tokenizerBosToken, tokenizerEosToken, properties->tokenizer.get_bos_token_id(), properties->tokenizer.get_eos_token_id(), isGGUFModel, tokenizerTemplate);
+
+                std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
+                SPDLOG_TRACE("Time to get bos/eos tokens from tokenizer: {} ms", std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() / 1000.0);
+            }
+
+            if (tokenizerEosToken.empty()) {
+                SPDLOG_ERROR("Tokenizer eos token not found in tokenizer nor in vocabulary but required for GGUF models.");
+                throw std::runtime_error("Tokenizer eos token not found in tokenizer nor in vocabulary but required for GGUF models.");
+            } else {
+                properties->ggufEosToken = tokenizerEosToken;
+            }
+        }
+
+        auto locals = py::dict("tokenizer_template"_a = tokenizerTemplate,
+            "tokenizer_bos_token"_a = tokenizerBosToken,
+            "tokenizer_eos_token"_a = tokenizerEosToken,
+            "templates_directory"_a = chatTemplateDirectory,
+            "is_gguf_model"_a = isGGUFModel);
```

Comment:
Can  you extract that to another method? Maybe we could have a struct `ExtraGenerationInfo` that would hold `bosTokenFromTokenizer`, `bosTokenIdFromTokenizer`, `eosTokenFromTokenizer`, `eosTokenIdFromTokenizer`, `chatTemplateFromTokenizer`, `isGgufModel`. 
That new method would read that data and create `ExtraGenerationInfo` that we could then pass to this `loadPyTemplateProcessor` and avoid tokenizer dependency here. 

So we would have something like:
```
extraGenerationInfo = readExtraGenerationInfo(modelDirectory, tokenizer)
loadPyTemplateProcessor(properties, extraGenerationInfo)
```
in ServableInitializer::initialize() method.